### ISSUE TITLE
Dracut throws an error for builtin modules

### ISSUE THREAD

**Issue by ddiss** [rapido-linux/rapido#130]
I recently upgraded to dracut-049.1 and found that the `samba-local` rapido image no longer works. This appears to be due to Dracut no longer gracefully handling `--add-drivers` entries for builtin kernel modules. I think this should probably be an upstream Dracut report, but I'd like to do some debugging here first.

**Comment by ddiss** [comment]
Given that dracut remains broken here, I'm considering just passing through the entire `KERNEL_INSTALL_MOD_PATH` as a readonly virtfs/9p share. Such a setup would mean that virtio/9p support must be built-in, but I think that's already the case. Another question would be how running VMs behave if the modules are modified or overwritten.

**Comment by ddiss** [comment]
> Given that dracut remains broken here, I'm considering just passing through the entire `KERNEL_INSTALL_MOD_PATH` as a readonly virtfs/9p share. Such a setup would mean that virtio/9p support must be built-in, but I think that's already the case. 

This seems to be working okay and saves some time during cut, e.g.

fstests-btrfs, many modules
----
before:
```
> time ./rapido cut -B fstests-btrfs
...
real    0m4.325s
user    0m2.592s
sys     0m1.931s
```

after
```
> time ./rapido cut -B fstests-btrfs
...
real    0m3.686s
user    0m2.565s
sys     0m1.635s
```

simple-example, few modules
--
before
```
> time ./rapido cut -B simple-example
...
real    0m0.962s
user    0m0.752s
sys     0m0.341s
```

after:
```
> time ./rapido cut -B simple-example
...
real    0m0.946s
user    0m0.736s
sys     0m0.342s
```
I still need to do some boot time benchmarks.

> Another question would be how running VMs behave if the modules are modified or overwritten.

I think it's probably reasonable to leave this up to the responsibility of the user. i.e. kernel module rebuilds+installs shouldn't occur while a rapido VM may be reading the modules.

@morbidrsa and @luis-henrix : curious to hear what you guys think. I'll submit the change as an RFC patch.

**Comment by ddiss** [comment]
I did some quick `fstests-btrfs` tests with `FSTESTS_AUTORUN_CMD="shutdown"`.

before:
```
> time rapido boot
...
real    0m1.405s
user    0m1.248s
sys     0m0.267s
```

after:
```
> time rapido boot
...
real    0m1.609s
user    0m1.412s
sys     0m0.334s
```

So the `cut` time saving needs to be weighed up against the increase in boot time due to virtfs mount + proxied module reads.

> I'll submit the change as an RFC patch.

I pushed to https://github.com/ddiss/rapido/tree/kmods_via_virtfs but I'm now starting to doubt this change... One feature I'd hoped to add at some stage was the ability to feed kernel and initramfs to other (cloud) hypervisors or physical machines via pxe/tftp, etc. Relying on virtfs for modules would break this dream ;-)

**Comment by luis-henrix** [comment]
I'll be honest: I don't really like this workaround.

So, the problem is that, when using `--add-drivers` with a driver that doesn't really exist (which may or may not be built-in), dracut emits an error and doesn't add any of the other kernel modules on the the list.  For example, if I use `--add-drivers "zram lzo foo lzo-rle"` my initrd will not contain the `lzo-rle` module because `foo` doesn't exist.

IMO, the best options are:
1. Fix dracut.  Although I'm not really sure it's "broken", maybe that's really the correct behaviour.
2. Fix the cut scripts, removing the extra modules
3. Add an helper function in runtime.vars that would sanitize the drivers list before passing it to dracut.

**Comment by ddiss** [comment]
Thanks for the feedback, Luis.

> I'll be honest: I don't really like this workaround.
> 
> So, the problem is that, when using `--add-drivers` with a driver that doesn't really exist (which may or may not be built-in), dracut emits an error and doesn't add any of the other kernel modules on the the list. For example, if I use `--add-drivers "zram lzo foo lzo-rle"` my initrd will not contain the `lzo-rle` module because `foo` doesn't exist.

Exactly.

> IMO, the best options are:
> 
>     1. Fix dracut.  Although I'm not really sure it's "broken", maybe that's really the correct behaviour.

Yeah, I think this is the right way forward. I'm (almost) certain that it used to correctly parse `modules.builtin` / `modules.dep` so I'll have a crack at trying to fix this regression. 

>     2. Fix the cut scripts, removing the extra modules
> 
>     3. Add an helper function in runtime.vars that would sanitize the drivers list before passing it to dracut.

4. tell dracut to always pack all kernel modules into the initramfs image
   - much larger images, much slower cut (and likely boot) times

5. drop dracut and switch to kernel/usr/gen_init_cpio or u-root
  - a bit drastic ;-)

**Comment by ddiss** [comment]
I noticed that this got fixed in upstream Dracut just recently via:
```
commit 2536a9eaffbc9cc14c85579a2f537d3f3a1d5659
Author: Harald Hoyer <harald@redhat.com>
Date:   Fri Mar 19 10:59:21 2021 +0100

    fix(install): handle builtin modules
    
    If a `kmod_module` is missing the `path`, it is `builtin`.
```

It doesn't look like it's made it into openSUSE Leap 15.3 yet. I'll raise a downstream bug and submit there.
Looks like we can drop the `KERNEL_INSTALL_MOD_PATH` experiments. \o/

**Comment by ddiss** [comment]
> I noticed that this got fixed in upstream Dracut just recently via:
> commit 2536a9eaffbc9cc14c85579a2f537d3f3a1d5659
...
> It doesn't look like it's made it into openSUSE Leap 15.3 yet. I'll raise a downstream bug and submit there.

The fix has been merged for 15.3 via https://github.com/openSUSE/dracut/pull/124
